spec.bs

```diff
@@ -201,121 +216,325 @@ This document introduces a new storage API that is intentionally not partitioned
+        1. [=Assert=]: |operation|'s [=associated realm=] is [=this=]'s [=relevant realm=].
+        1. Let |argumentsList| be the [=/list=] « |urlList| ».
+        1. If |options| [=map/contains=] |data|, [=list/append=] |data| to |argumentsList|.
+        1. Let |operationResult| be the result of [=invoking=] |operation| with |argumentsList|.
+        1. If [=an exception was thrown=], then [=queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
+
+            Note: This indicates that |operationCtor|'s run() method encounters an error, where |operationCtor| is the parameter in {{SharedStorageWorkletGlobalScope/register()}}.
+
+        1. Otherwise:
+            1. Set |index| to |operationResult|.
+            1. If [=an exception was thrown=]:
+                1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
+                1. Abort these steps.
+
+                Note: The result index is a non-integer value. This violates the selectURL() protocol, and we don't know which url should be selected.
+
+            1. Otherwise, if |index| is greater than |urlList|'s [=list/size=]:
+                1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=reject=] |promise| with a {{TypeError}}.
+                1. Abort these steps.
+
+                Note: The result index is beyond the input urls' size. This violates the selectURL() protocol, and we don't know which url should be selected.
+
+            1. [=Queue a global task=] on the [=DOM manipulation task source=], given |window|, to [=resolve=] |promise| with |index|.
+  </div>
+
+  <div algorithm>
+    The <dfn method for="SharedStorageWorklet">selectURL(|name|, |urls|, |options|)</dfn> method steps are:
+
+    1. Let |resultPromise| be a new [=promise=].
+    1. If [=this=]'s [=addModule initiated=] is false, then return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |window| be [=this=]'s [=relevant settings object=].
+    1. [=Assert=]: |window| is a {{Window}}.
+    1. Let |context| be |window|'s [=Window/browsing context=].
+    1. If |context| is null, then return a [=promise rejected=] with a {{TypeError}}.
+    1. Let |document| be |context|'s [=active document=].
+    1. [=Assert=]: [=this=]'s [=global scopes=]'s [=list/size=] is 1.
+    1. Let |globalScope| be [=this=]'s [=global scopes=][0].
+    1. Let |workletOrigin| be |globalScope|'s [=global object/realm=]'s \[[HostDefined]] field's [=environment settings object/origin=].
+    1. If the result of running [=Is feature enabled in document for origin?=] on "[=PermissionsPolicy/shared-storage-select-url=]", |document|, and |workletOrigin| returns false, return a [=promise rejected=] with a {{TypeError}}.
+    1. If [=this=]'s [=global scopes=] is [=list/empty=], then return a [=promise rejected=] with a {{TypeError}}.
```

Comment:
Fixed. Although, I believe it should be `This can happen if either {{WindowSharedStorage/selectURL()}} or {{SharedStorageWorklet/selectURL()}} is called before {{addModule()}}`. This isn't applicable to createWorklet(), because createWorklet() will create the global scope and the (new) worklet object at the same time, and thus it's guaranteed that selectURL() is called after that.

Reply:
Ah, you're saying selectURL() _cannot_ be called after createWorklet() then, right?

Reply:
selectURL() _can_ be called after createWorklet(). Not sure which selectURL() you are referring to, but:
- `sharedStorage.selectURL()` will defer to `sharedStorage.worklet.selectURL()`
- For an explicitly created SharedStorageWorklet object (via `let worklet = await sharedStorage.createWorklet()`, the new worklet object should also be able to call `worklet.selectURL()` after creation.

Reply:
Oops I'm sorry. I meant "_before_" in my previous comment.

Reply:
It'd be an ambiguous statement. `createWorklet()` operates on `window.sharedStorage`, and its _result_ is a `SharedStorageWorklet` object, which can further invoke `worklet.selectURL()`. But if you are talking about `sharedStorage.createWorklet()` v.s `sharedStorage.selectURL()`, they are _unrelated_, because they create/operate on different worklet object.